sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java

```diff
@@ -0,0 +1,228 @@
+        private FieldProvider<Symbol> inputColumnProvider = new FieldProvider<Symbol>() {
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, Operation operation) {
+                return resolveField(qualifiedName, null, operation);
+            }
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, @Nullable List<String> path, Operation operation) {
+                return new InputColumn(0);
+            }
+        };
+
+        TestIngestionImplementation(Functions functions,
+                                    IngestionService ingestionService) {
+            this.ingestRules = new HashSet<>();
+            this.inputFactory = new InputFactory(functions);
+            this.ingestionService = ingestionService;
+            this.expressionAnalysisContext = new ExpressionAnalysisContext();
+            this.expressionAnalyzer = new ExpressionAnalyzer(functions, null, null, inputColumnProvider, null);
+        }
+
+        void registerIngestionImplementation() {
+            ingestionService.registerImplementation(INGESTION_SOURCE_ID, this);
+        }
+
+        void ingestData(String topic, int data) {
+            if (ingestRules.size() > 0) {
+                for (IngestRule ingestRule : ingestRules) {
+                    Symbol conditionSymbol = expressionAnalyzer.convert(SqlParser.createExpression(ingestRule.getCondition()), expressionAnalysisContext);
+                    Predicate<Row> conditionPredicate = RowFilter.create(inputFactory, conditionSymbol);
+                    if (conditionPredicate.test(new Row1(topic))) {
+                        execute(String.format(Locale.ENGLISH,
+                            "insert into %s values (%s);", ingestRule.getTargetTable(), data)
+                        );
+                        execute("refresh table " + ingestRule.getTargetTable());
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void applyRules(Set<IngestRule> rules) {
+            this.ingestRules = rules;
+        }
+    }
+
+    private TestIngestionImplementation implementation;
+    private Thread dataIngestionThread;
+    private Iterator<Integer> dataSource = Stream.iterate(1, i -> i + 1).iterator();
+
+    private volatile boolean produceData;
+    private CountDownLatch dataFlowingLatch;
+    private final AtomicInteger lastProducedData = new AtomicInteger(-1);
+    private final Semaphore dataFlowSemaphore = new Semaphore(1);
+
+    private void pauseDataIngestion() throws InterruptedException {
+        dataFlowSemaphore.acquire();
+    }
+
+    private void resumeDataIngestion() {
+        dataFlowSemaphore.release();
+    }
+
+    @Before
+    public void setupIngestRuleAndImplementation() {
+        execute("create table ingest_data_raw (data int)");
+        execute(String.format(Locale.ENGLISH,
+            "create ingest rule %s on %s where topic = '%s' into %s",
```

Comment:
would prefer using parameterSubstitution wherever possible (avoiding people would copy this pattern ;)